### PR TITLE
core/simple-stream: add .position() method to input streams

### PR DESCRIPTION
**Pull request by enedil**
Position is needed if we want to partition the stream into a list of substreams, where substream sizes are encoded in the stream itself. Then we can decode the stream before and after reading next part (that would comprise next substream), and generate a substream of the computed length.

This is needed for implementation of https://github.com/scylladb/scylladb/issues/11816

CC @bhalevy

### CODE REVIEW

File: include/seastar/core/simple-stream.hh [hunk 1]
```diff
@@ -148,6 +149,7 @@ public:
     void skip(size_t size) {
         for_each_fragment(size, [] (auto) { });
     }
+
```
Comment (bhalevy) [inline include/seastar/core/simple-stream.hh:152]: nit: please refrain from mixing cosmetic changes with functional changes.
In this case, the cosmetic changes don't seem to be required at all...

File: include/seastar/core/simple-stream.hh [hunk 2]
```diff
@@ -367,6 +370,11 @@ public:
     size_t size() const {
         return _size;
     }
+
+    [[gnu::always_inline]]
+    size_t position() const {
```
Comment (bhalevy) [inline include/seastar/core/simple-stream.hh:375]: nit: `const noexcept` please

File: include/seastar/core/simple-stream.hh [hunk 3]
```diff
@@ -396,15 +405,18 @@ private:
...
     friend class fragmented_memory_output_stream<Iterator>;
 public:
     using has_with_stream = std::false_type;
     using iterator_type = Iterator;
     fragmented_memory_input_stream(Iterator it, size_t size)
-        : _it(it), _size(size) {
+        : _it(it), _size(size), _initial_size(size) {
     }
 
+    size_t position() const {
```
Comment (bhalevy) [inline include/seastar/core/simple-stream.hh:417]: nit: `const noexcept`
why any reason for not marking this one with `[[gnu::always_inline]]` too?

Comment (nyh) [reply include/seastar/core/simple-stream.hh:417]: @bhalevy why add this ugly and non-standard `[[gnu::always_inline]]``? What does it add?

**Review by nyh — CHANGES_REQUESTED** [review]
Hi, thanks. your implementation seems reasonable, but I don't understand the motivation...

First, the specific example you gave seems strange specific and doesn't even explain why position() is needed for this specific use case.

Moreover, it seems to me that position() not really needed in that specific use case... The method that carves out a "substream" from a stream could start at **the current position** (without knowing or caring exactly that is position 12345 in the original stream) and **a given length forward**. In fact, we already have such a feature in Seastar's HTTP implementation which needs to carve the large input stream into pieces of individual request bodies. As you can see it doesn't need position().

Finally, I must have been sleeping under a rock, but I have to admit I never noticed this "simple-stream" before. Is this a type of input_stream? If so, why would we want to add the position() API only for simple-stream but not for the more general input-stream?

File: include/seastar/core/simple-stream.hh (hunk 3 above)
Comment (bhalevy) [reply include/seastar/core/simple-stream.hh:417]: Without judging if it's helpful or not,
we need to add to add it in both cases or neither.
I think the compiler will inline it anyway when it's defined as a method in the header file so I agree it's not needed,
but that said it's already present in this header.  I wonder why

**Comment by enedil** [comment]
After all I could avoid adding that .position parameter, so it wasn't needed :P 
The length forward was just not know, so initially I opted for this easier way. But it turns out that give .size() before deserialization and .size() after can give me the length. So this PR can be discarded.

**Pull request closed without merging**